runtime/src/test/java/org/corfudb/runtime/ClientHandshakeHandlerTest.java

```diff
@@ -0,0 +1,190 @@
+import org.corfudb.runtime.proto.service.CorfuMessage.ProtocolVersionMsg;
+import org.corfudb.runtime.proto.service.CorfuMessage.RequestMsg;
+import org.corfudb.runtime.proto.service.CorfuMessage.ResponseMsg;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.DEFAULT_UUID;
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolBase.getHandshakeResponseMsg;
+import static org.corfudb.protocols.service.CorfuProtocolBase.getPingResponseMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getHeaderMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getResponseMsg;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class ClientHandshakeHandlerTest {
+
+    // Netty EmbeddedChannel for unit testing.
+    private EmbeddedChannel embeddedChannel;
+
+    // Testing appender for collecting log messages during test execution.
+    private ListAppender<ILoggingEvent> listAppender;
+
+    // Some utility variables needed for following test cases.
+    private final AtomicInteger requestCounter = new AtomicInteger();
+
+    private static final int DEFAULT_HANDSHAKE_TIMEOUT = 10;
+
+    private static final long FAKE_SERVER_VERSION = 111111111;
+
+    private static final UUID SERVER_NODEID = UUID.randomUUID();
+
+    /**
+     * A helper method that creates a basic message header populated
+     * with default values.
+     *
+     * @param ignoreClusterId   indicates if the message is clusterId aware
+     * @param ignoreEpoch       indicates if the message is epoch aware
+     * @return                  the corresponding HeaderMsg
+     */
+    private HeaderMsg getBasicHeader(ClusterIdCheck ignoreClusterId, EpochCheck ignoreEpoch) {
+        return getHeaderMsg(requestCounter.incrementAndGet(), PriorityLevel.NORMAL, 0L,
+                getUuidMsg(DEFAULT_UUID), getUuidMsg(DEFAULT_UUID), ignoreClusterId, ignoreEpoch);
+    }
+
+    @Before
+    public void setup() {
+        ClientHandshakeHandler clientHandshakeHandler = new ClientHandshakeHandler(DEFAULT_UUID,
+                SERVER_NODEID, DEFAULT_HANDSHAKE_TIMEOUT);
+        embeddedChannel = new EmbeddedChannel(clientHandshakeHandler);
+
+        // Setting up the logger such that we could inspect the logged messages in ClientHandshakeHandler.
+        Logger logger = (Logger) LoggerFactory.getLogger(ClientHandshakeHandler.class);
+        logger.setLevel(Level.DEBUG);
+
+        listAppender = new ListAppender<>();
+        listAppender.start();
+        logger.addAppender(listAppender);
```

Comment:
I think relying on logging for unit tests makes them fragile. Handshake success/failure can be tested by detecting the presence or absence of another message sent down the pipeline. In the case of a version mismatch, maybe there isn't enough substance for a unit test, since it has no impact on the behaviour yet.

Reply:
Changed, thanks!